cli/flox-rust-sdk/src/providers/services.rs

```diff
@@ -367,9 +372,173 @@ impl From<ProcessComposeLogContents> for LoggedError {
+                        // we can continue to the next reader
+                        Ok(()) => continue,
+                        Err(e) => return Some(Err(e)),
+                    }
+                }
+            },
+        }
+    }
+}
+
+/// Representation of a thread reading logs from a `process-compose process logs` process.
+struct ProcessComposeLogReader {
+    handle: std::thread::JoinHandle<Result<(), ServiceError>>,
+}
+
+impl ProcessComposeLogReader {
+    fn start(
+        socket: impl AsRef<Path>,
+        process: impl AsRef<str>,
+        tx: Sender<ProcessComposeLogLine>,
+    ) -> Result<ProcessComposeLogReader, ServiceError> {
+        let socket = socket.as_ref().to_path_buf();
+        let process = process.as_ref().to_string();
+
+        let handle = std::thread::spawn(move || {
+            let span = tracing::debug_span!("process-compose-log-reader", process = &process);
+            let _guard = span.enter();
+
+            let mut cmd = base_process_compose_command(socket);
+            cmd.arg("logs")
+                .arg(&process)
+                .arg("--follow")
+                .stderr(Stdio::piped())
+                .stdout(Stdio::piped());
+
+            debug!(cmd = cmd.display().to_string(), "attaching to logs");
+
+            let mut child = cmd.spawn().map_err(ServiceError::ProcessComposeCmd)?;
+
+            let stdout = child.stdout.take().expect("failed to get stdout");
+            let reader = std::io::BufReader::new(stdout);
+
+            // `process-compose process logs` will keep blocking even
+            // when the process **and socket** are gone.
+            // Thus reader.lines() will keep blocking indefinitely.
+            // Todo: add a heartbeat to stop receiving logs and kill log reader processes
+            for line in reader.lines() {
+                let line = line.map_err(ServiceError::ReadLogLine)?;
+
+                // The receiver end was dropped, so we can't send any more messages.
+                // Might as well break out of the loop and kill the child process.
+                let Ok(_) = tx.send(ProcessComposeLogLine::new(&process, line)) else {
+                    debug!("receiver dropped, stopping log reader");
+                    break;
+                };
+            }
+
+            // Here, either sending to the receiver failed i.e. the receiver was dropped,
+            // or the child process died.
+            // If the child process ended, that's unexpected
+            // and we'll try to communicate that through the channel.
+            // The most likely error is that the socket doesn't exist,
+            // trying to read logs for a non existent process
+            // unfortunately just blocks indefinitely without any error message.
+
+            if let Some(exit_status) = child.try_wait().map_err(ServiceError::ProcessComposeCmd)? {
+                debug!(exit_status = ?exit_status, "child process exited");
```

Comment:
```suggestion
                debug!(?exit_status, "child process exited");
```

Reply:
Also probably a good idea to list which process's logs this was reading

Reply:
the process is already on the span this is on, would you include it here again?